datastore/src/main/java/org/jboss/pnc/datastore/repositories/ProductMilestoneRepositoryImpl.java

```diff
@@ -30,4 +54,194 @@ public class ProductMilestoneRepositoryImpl extends AbstractRepository<ProductMi
+        CriteriaQuery<Long> query = cb.createQuery(Long.class);
+
+        Root<ProductMilestone> milestones = query.from(ProductMilestone.class);
+        SetJoin<ProductMilestone, Artifact> deliveredArtifacts = milestones.join(ProductMilestone_.deliveredArtifacts);
+        Path<ProductMilestone> builtDeliveredArtifactsMilestone = deliveredArtifacts.get(Artifact_.buildRecord)
+                .get(BuildRecord_.productMilestone);
+
+        // 1) we care about artifacts from the milestone given by id
+        // 2) only built delivered artifacts *not* from this product
+        query.where(
+                cb.equal(milestones.get(ProductMilestone_.id), id),
+                cb.notEqual(
+                        builtDeliveredArtifactsMilestone.get(ProductMilestone_.productVersion)
+                                .get(ProductVersion_.product)
+                                .get(Product_.id),
+                        getProductIdByItsMilestone(id)));
+        query.select(cb.count(deliveredArtifacts.get(Artifact_.id)));
+
+        return entityManager.createQuery(query).getSingleResult();
+    }
+
+    @Override
+    public long countDeliveredArtifactsBuiltInNoMilestone(Integer id) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Long> query = cb.createQuery(Long.class);
+
+        Root<ProductMilestone> milestones = query.from(ProductMilestone.class);
+        SetJoin<ProductMilestone, Artifact> deliveredArtifacts = milestones.join(ProductMilestone_.deliveredArtifacts);
+        Join<Artifact, BuildRecord> deliveredArtifactsBuild = deliveredArtifacts.join(Artifact_.buildRecord);
+
+        // 1) we care about artifacts from the milestone given by id
+        // 2) only built delivered artifacts with no milestone
+        query.where(
+                cb.equal(milestones.get(ProductMilestone_.id), id),
+                cb.isNull(deliveredArtifactsBuild.get(BuildRecord_.productMilestone)));
+        query.select(cb.count(deliveredArtifacts.get(Artifact_.id)));
+
+        return entityManager.createQuery(query).getSingleResult();
+    }
+
+    @Override
+    public long countDeliveredArtifactsNotBuilt(Integer id) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Long> query = cb.createQuery(Long.class);
+
+        Root<ProductMilestone> milestones = query.from(ProductMilestone.class);
+        SetJoin<ProductMilestone, Artifact> deliveredArtifacts = milestones.join(ProductMilestone_.deliveredArtifacts);
+        Join<Artifact, BuildRecord> deliveredArtifactsBuild = deliveredArtifacts
+                .join(Artifact_.buildRecord, JoinType.LEFT);
+
+        // 1) we care about artifacts from the milestone given by id
+        // 2) only delivered artifacts which were *not* built
+        query.where(
+                cb.equal(milestones.get(ProductMilestone_.id), id),
+                cb.isNull(deliveredArtifactsBuild.get(BuildRecord_.id)));
```

Comment:
What about this?
`cb.isNull(deliveredArtifacts.get(Artifact_.buildRecord))`

Reply:
Nice catch Honza, that _deliveredArtifactsBuild_ was not needed at all